host/retry_policy_workflow_test.go

```diff
@@ -0,0 +1,290 @@
+
+	var history *types.History
+	var lastEvent *types.HistoryEvent
+
+GetHistoryLoop:
+	for i := 0; i < 20; i++ {
+		historyResponse, err := s.engine.GetWorkflowExecutionHistory(createContext(), &types.GetWorkflowExecutionHistoryRequest{
+			Domain: s.domainName,
+			Execution: &types.WorkflowExecution{
+				WorkflowID: id,
+			},
+		})
+		s.Nil(err)
+		history = historyResponse.History
+
+		lastEvent = history.Events[len(history.Events)-1]
+		if *lastEvent.EventType != types.EventTypeWorkflowExecutionTimedOut {
+			s.Logger.Warn("Execution not timedout yet.")
+			time.Sleep(200 * time.Millisecond)
+			continue GetHistoryLoop
+		}
+
+		timeoutEventAttributes := lastEvent.WorkflowExecutionTimedOutEventAttributes
+		s.Equal(types.TimeoutTypeStartToClose, *timeoutEventAttributes.TimeoutType)
+		break GetHistoryLoop
+	}
+
+	firstEvent := history.Events[0]
+	s.Equal(firstEvent.WorkflowExecutionStartedEventAttributes.Attempt, int32(1))
+
+	delta := time.Unix(0, common.Int64Default(firstEvent.WorkflowExecutionStartedEventAttributes.ExpirationTimestamp)).Sub(now)
+	s.True(delta > 2*time.Second)
+	s.True(delta < 4*time.Second)
+}
+
+func (s *IntegrationSuite) TestWorkflowRetryPolicyContinueAsNewAsCron() {
+	id := "integration-workflow-retry-policy-continue-as-new-cron-test"
+	wt := "integration-workflow-retry-policy-continue-as-new-cron-test-type"
+	tl := "integration-workflow-retry-policy-continue-as-new-cron-test-tasklist"
+	identity := "worker1"
+
+	workflowType := &types.WorkflowType{}
+	workflowType.Name = wt
+
+	taskList := &types.TaskList{}
+	taskList.Name = tl
+
+	retryPolicy := &types.RetryPolicy{
+		InitialIntervalInSeconds:    1,
+		BackoffCoefficient:          2.0,
+		MaximumIntervalInSeconds:    1,
+		ExpirationIntervalInSeconds: 1,
+		MaximumAttempts:             0,
+		NonRetriableErrorReasons:    []string{"bad-error"},
+	}
+	request := &types.StartWorkflowExecutionRequest{
+		RequestID:                           uuid.New(),
+		Domain:                              s.domainName,
+		WorkflowID:                          id,
+		WorkflowType:                        workflowType,
+		TaskList:                            taskList,
+		Input:                               nil,
+		ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(3),
+		TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(3),
+		Identity:                            identity,
+		RetryPolicy:                         retryPolicy,
+	}
+
+	we, err0 := s.engine.StartWorkflowExecution(createContext(), request)
+	s.Nil(err0)
+
+	s.Logger.Info("StartWorkflowExecution", tag.WorkflowRunID(we.RunID))
+	dtHandler := func(execution *types.WorkflowExecution, wt *types.WorkflowType,
+		previousStartedEventID, startedEventID int64, history *types.History) ([]byte, []*types.Decision, error) {
+		buf := new(bytes.Buffer)
+		return []byte(strconv.Itoa(0)), []*types.Decision{{
+			DecisionType: types.DecisionTypeContinueAsNewWorkflowExecution.Ptr(),
```

Comment:
similar to retry policy, continue as new for cron is also create by server, decision task handler is only needs to return workflow completed here. In start workflow request, you'll need to specify a cron schedule. like `@every 3s` and you may need to recalculate the sleep time below.